group/edwards25519/scalar.go

```diff
@@ -442,6 +627,1286 @@ func scMulAdd(s, a, b, c *[32]byte) {
 	s[31] = byte(s11 >> 17)
 }
 
+// Hacky scAdd cobbled together rather sub-optimally from scMulAdd.
+//
+// Input:
+//   a[0]+256*a[1]+...+256^31*a[31] = a
+//   c[0]+256*c[1]+...+256^31*c[31] = c
+//
+// Output:
+//   s[0]+256*s[1]+...+256^31*s[31] = (a+c) mod l
+//   where l = 2^252 + 27742317777372353535851937790883648493.
+//
+func scAdd(s, a, c *[32]byte) {
+	a0 := 2097151 & load3(a[:])
+	a1 := 2097151 & (load4(a[2:]) >> 5)
+	a2 := 2097151 & (load3(a[5:]) >> 2)
+	a3 := 2097151 & (load4(a[7:]) >> 7)
+	a4 := 2097151 & (load4(a[10:]) >> 4)
+	a5 := 2097151 & (load3(a[13:]) >> 1)
+	a6 := 2097151 & (load4(a[15:]) >> 6)
+	a7 := 2097151 & (load3(a[18:]) >> 3)
+	a8 := 2097151 & load3(a[21:])
+	a9 := 2097151 & (load4(a[23:]) >> 5)
+	a10 := 2097151 & (load3(a[26:]) >> 2)
+	a11 := (load4(a[28:]) >> 7)
+	c0 := 2097151 & load3(c[:])
+	c1 := 2097151 & (load4(c[2:]) >> 5)
+	c2 := 2097151 & (load3(c[5:]) >> 2)
+	c3 := 2097151 & (load4(c[7:]) >> 7)
+	c4 := 2097151 & (load4(c[10:]) >> 4)
+	c5 := 2097151 & (load3(c[13:]) >> 1)
+	c6 := 2097151 & (load4(c[15:]) >> 6)
+	c7 := 2097151 & (load3(c[18:]) >> 3)
+	c8 := 2097151 & load3(c[21:])
+	c9 := 2097151 & (load4(c[23:]) >> 5)
+	c10 := 2097151 & (load3(c[26:]) >> 2)
+	c11 := (load4(c[28:]) >> 7)
+	var carry [23]int64
+
+	s0 := c0 + a0
+	s1 := c1 + a1
+	s2 := c2 + a2
+	s3 := c3 + a3
+	s4 := c4 + a4
+	s5 := c5 + a5
+	s6 := c6 + a6
+	s7 := c7 + a7
+	s8 := c8 + a8
+	s9 := c9 + a9
+	s10 := c10 + a10
+	s11 := c11 + a11
+	s12 := int64(0)
+	s13 := int64(0)
+	s14 := int64(0)
+	s15 := int64(0)
+	s16 := int64(0)
+	s17 := int64(0)
+	s18 := int64(0)
+	s19 := int64(0)
+	s20 := int64(0)
+	s21 := int64(0)
+	s22 := int64(0)
+	s23 := int64(0)
+
```

Comment:
@bford Indeed you were completely right !
```
BenchmarkCTScalarAdd-4                  20000000               104 ns/op
BenchmarkCTScalarSimpleAdd-4            10000000               196 ns/op
BenchmarkCTScalarFactoredAdd-4           2000000               592 ns/op
BenchmarkCTScalarMul-4                  10000000               166 ns/op
BenchmarkCTScalarSimpleMul-4            10000000               200 ns/op
BenchmarkCTScalarFactoredMul-4           2000000               722 ns/op
BenchmarkCTScalarSub-4                  20000000                98.5 ns/op
BenchmarkCTScalarSimpleSub-4            10000000               192 ns/op
BenchmarkCTScalarFactoredSub-4           2000000               610 ns/op
```
where `ScalarFactoredXXX` is is using a factored out version of `scMul` etc.
Crazy thought the big difference in time. I'd love to see this kind of benchmark on Rust as well... And also to take some compiler courses to really grasp the thing > < :)